source/agora/common/MerkleTree.d

```diff
@@ -0,0 +1,233 @@
+
+    public Hash getHash ()
+        nothrow @nogc @trusted
+    {
+        if (this.root is null)
+            return Hash.init;
+        else
+            return this.root.getHash();
+    }
+}
+
+/// The `MerkleNode` has one transaction or two `MerkleNode`
+public struct MerkleNode
+{
+    /// The left branch's ptr of the tree.
+    public MerkleNode* left;
+
+    /// The right branch's ptr of the tree.
+    public MerkleNode* right;
+
+    /// The transaction's ptr
+    public Transaction* tx;
+
+    /// The hash of this `MerkleNode`
+    public Hash hash;
+
+    /***************************************************************************
+
+        Create
+
+        Params:
+            left = The left branch's ptr of the tree.
+            right = The right branch's ptr of the tree.
+            tx = The transaction's ptr
+
+    ***************************************************************************/
+
+    /// Ctor
+    this (MerkleNode* left, MerkleNode* right, Transaction* tx = null)
+    {
+        this.left = left;
+        this.right = right;
+        this.tx = tx;
+        this.hash = this.getHash();
+    }
+
+    /***************************************************************************
+
+        Calculate the hash value of the `MerkleNode`
+
+        Return:
+            The `Hash` representing this instance
+
+    ***************************************************************************/
+
+    public Hash getHash ()
+        nothrow @nogc @trusted
+    {
+        if ((this.left is null) && (this.right is null))
+        {
+            if (this.tx !is null)
+                return hashFull(*this.tx);
+            else
+                return Hash.init;
+        }
+        else if ((this.left !is null) && (this.right !is null))
+        {
+            return mergeHash(this.left.hash, this.right.hash);
+        }
+        else
+            return Hash.init;
+    }
+}
+
+
+/*******************************************************************************
+
+    Found nodes with specific hash on child nodes.
+
+    Params:
+        parent = parent node.
+        hash = looking hash.
+
+    Returns:
+        Return node if found the hash. Return null otherwise.
+
+*******************************************************************************/
+
+public MerkleNode* searchNode (MerkleNode* parent, Hash hash)
+{
+    if (parent is null) return null;
+
+    if (parent.hash == hash) return parent;
+
+    MerkleNode* res;
+    res = searchNode(parent.right, hash);
+    if (res != null) return res;
+
+    res = searchNode(parent.left, hash);
+    if (res != null) return res;
+
+    return null;
+}
+
+/// Merge two hash
+public Hash mergeHash (Hash h1, Hash h2)
+        nothrow @nogc @trusted
+{
+    static struct MergeHash
+    {
+        Hash left;
+        Hash right;
+
+        public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+        {
+            if (this.left < this.right)
```

Comment:
Why ?

Reply:
It is to change the order of the two so that the same value can be obtained.
We compared the values first, and we got the hash value ahead of the small one.

Reply:
It however means you need to change the order of the transaction at the lower layer.